runtime/compiler/p/codegen/J9CodeGenerator.cpp

```diff
@@ -403,7 +403,7 @@ J9::Power::CodeGenerator::insertPrefetchIfNecessary(TR::Node *node, TR::Register
    static bool disableStringObjPrefetch = (feGetEnv("TR_DisableStringObjPrefetch") != NULL);
    bool optDisabled = false;
 
-   if (node->getOpCodeValue() == TR::aloadi ||
+   if ((node->getOpCodeValue() == TR::aloadi && !comp->target().is64Bit()) ||
         (comp->target().is64Bit() &&
```

Comment:
This looks like conditions being misplaced originally.  This current fix certainly is not in line with the intention of prefetch optimization.  Prefetch surely is beneficial performance-wise to 64bit non-compressed.

The conditions should read something like:   if  (is-objectReference-load  &&  hotness-testing && CPU-testing)

Applies to the next change too.

Reply:
While I agree that prefetching certainly could be applied in these circumstances, extending the prefetching optimization to handle this case correctly is out-of-scope for this PR. As the commit message clarifies, the previous implementation was not performing prefetching for such nodes either, but the way this was being achieved was slightly different. This change does not make prefetching not occur where it previously would, but simply makes `insertPrefetchIfNecessary` safe to call on `aloadi` nodes under 64-bit, as it would otherwise segfault.

Previously, we were simply failing to call `insertPrefetchIfNecessary` at all under the problematic conditions, thus sidestepping the problem. In commoning up code between the load evaluators into `LoadStoreHandler`, all of these evaluators need to behave uniformly and thus now call `insertPrefetchIfNecessary` unconditionally, with it being up to that method alone to make the decision as to whether to perform prefetching. This also avoids leaking implementation details of OpenJ9's prefetching into OMR.

If you'd like, I can open an issue for extending this support in the future, though.

Reply:
certainly need a follow-up item

Reply:
Opened #11803